File: src/main/java/com/converter/api/validation/UnregisteredUserEmailValidator.java
package com.converter.api.validation;

import com.converter.api.repository.UserRepository;

import javax.validation.ConstraintValidator;
import javax.validation.ConstraintValidatorContext;

public class UnregisteredUserEmailValidator implements ConstraintValidator<UnregisteredUserEmail, String> {

    private final UserRepository userRepository;

    public UnregisteredUserEmailValidator(UserRepository userRepository) {
        this.userRepository = userRepository;
    }

    @Override
    public boolean isValid(String email, ConstraintValidatorContext constraintValidatorContext) {
        if (email == null) {
            return true;
        } else {
            return !userRepository.existsByEmail(email);
        }
    }
}

File: src/main/kotlin/com/converter/api/validation/UnregisteredUserEmailValidator.kt
package com.converter.api.validation

import com.converter.api.repository.UserRepository
import javax.validation.ConstraintValidator
import javax.validation.ConstraintValidatorContext

class UnregisteredUserEmailValidator(
    private val userRepository: UserRepository
) : ConstraintValidator<UnregisteredUserEmail, String?> {

    override fun isValid(email: String?, constraintValidatorContext: ConstraintValidatorContext): Boolean {
        return if (email == null) {
            true
        } else {
            !userRepository.existsByEmail(email)
        }
    }
}